### ISSUE TITLE
Non-zero status code returned while running Resize node

### ISSUE THREAD

**Issue by adepierre** [microsoft/onnxruntime#13975]
### Describe the issue

Trying to run a ONNX model with a Resize layer with DML results in this error message:

```
[E:onnxruntime:, sequential_executor.cc:369 onnxruntime::SequentialExecutor::Execute] Non-zero status code returned while running Resize node. Name:'Resize_0' Status Message: D:\a\_work\1\s\onnxruntime\core\providers\dml\DmlExecutionProvider\src\MLOperatorAuthorImpl.cpp(1866)\onnxruntime.dll!00007FFFDD662A90: (caller: 00007FFFDD661F93) Exception(3) tid(3728) 80070057 ParamDone
```

### To reproduce

Model export with PyTorch :

```python
import torch
from torch.nn import functional as F

class TestModel(torch.nn.Module):
    def forward(self, x):
        return F.interpolate(x, scale_factor=2, mode='nearest')

model = TestModel()
example = torch.rand(1, 3, 256, 256)
torch.onnx.export(model,
                    example,
                    'test.onnx',
                    input_names=['inputs'],
                    export_params=True,
                    dynamic_axes={'inputs': {2 : 'height', 3: 'width'}}
)
```

Running the model in C++ (tested with [Microsoft.ML.OnnxRuntime.DirectML.1.13.1.zip](https://github.com/microsoft/onnxruntime/releases/download/v1.13.1/Microsoft.ML.OnnxRuntime.DirectML.1.13.1.zip))
```cpp
#include <iostream>
#include <vector>
#include "onnxruntime_cxx_api.h"
#include "dml_provider_factory.h"


int main(int argc, char** argv)
{
#ifdef _WIN32
    std::string str = "test.onnx";
    std::wstring wide_string = std::wstring(str.begin(), str.end());
    std::basic_string<ORTCHAR_T> model_file = std::basic_string<ORTCHAR_T>(wide_string);
#else
    std::string model_file = "test.onnx";
#endif

    Ort::Env env(ORT_LOGGING_LEVEL_WARNING, "test");
    Ort::SessionOptions session_options;
    OrtSessionOptionsAppendExecutionProvider_DML(session_options, 0);
    Ort::Session session = Ort::Session(env, model_file.data(), session_options);

    std::vector<int64_t> input_shape = { 1, 3, 196, 196 };
    std::vector<float> input_tensor_values(3 * 196 * 196, 0.0f);

    std::vector<Ort::Value> input_tensors;
    Ort::MemoryInfo memoryInfo = Ort::MemoryInfo::CreateCpu(OrtAllocatorType::OrtArenaAllocator, OrtMemType::OrtMemTypeDefault);
    input_tensors.push_back(Ort::Value::CreateTensor<float>(memoryInfo, input_tensor_values.data(), input_tensor_values.size(), input_shape.data(), input_shape.size()));

    std::vector<const char const*> input_names = { "inputs" };
    std::vector<const char const*> output_names = { "outputs" };

    std::vector<Ort::Value> outputs = session.Run(Ort::RunOptions{}, input_names.data(), input_tensors.data(), input_tensors.size(), output_names.data(), output_names.size());
    

    std::cout << outputs.size() << std::endl;
    std::cout << outputs[0].GetCount() << std::endl;
    return 0;
}
```

### Urgency

Can't infer model with resize layers in DML, so it's quite blocking

### Platform

Windows

### OS Version

10

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.13.1

### ONNX Runtime API

C++

### Architecture

X64

### Execution Provider

DirectML

**Comment by adepierre** [comment]
Quick update : I tried to replace the Resize op doing a quick "custom nearest interpolate function", but it leads to a similar issue, this time with Concat node 🙁
I think the root cause might be the dynamic size of the input, but I don't really know how to investigate more the issue and even less fix it

```python
import torch
from torch.nn import functional as F

def resize(x: torch.Tensor, scale_factor:int):
    stacked_h = torch.stack([x]*scale_factor, dim=-1).view(*x.shape[:-1], scale_factor*x.shape[-1])
    stacked = torch.stack([stacked_h]*scale_factor, dim=-2).view(*x.shape[:-2], scale_factor*x.shape[-2], scale_factor*x.shape[-1])

    return stacked

class TestModel(torch.nn.Module):
    def forward(self, x):
        return F.interpolate(x, scale_factor=2, mode='nearest')

model = TestModel()
example = torch.rand(1, 3, 256, 256)
torch.onnx.export(model,
                    example,
                    'test.onnx',
                    input_names=['inputs'],
                    export_params=True,
                    dynamic_axes={'inputs': {2 : 'height', 3: 'width'}}
)
```

```
[E:onnxruntime:, sequential_executor.cc:369 onnxruntime::SequentialExecutor::Execute] Non-zero status code returned while running Concat node. Name:'Concat_4' Status Message: D:\a\_work\1\s\onnxruntime\core\providers\dml\DmlExecutionProvider\src\MLOperatorAuthorImpl.cpp(1866)\onnxruntime.dll
```

**Comment by VladMVLX** [comment]
Unfortunately having the same problem now in April 2023 :(

**Comment by rodrigovimieiro** [comment]
Same problem now in November 2024

**Comment by make-a** [comment]
yes，i know this, hahaha, 

我再使用YOLO的动态输入时，遇到了一样的问题，我发现这是输入尺寸的问题，这可能同样适用于别的模型。

当使用**动态输入尺寸**的 YOLO11 模型时，如果 ONNX 模型被导出为动态尺寸，并直接使用输入图像的原始尺寸进行推理，可能会导致一些意外问题。  

通常，**输入图片的尺寸应该是某个特定数值的倍数**。YOLO 系列模型通常使用 **FPN（特征金字塔网络）** 或 **PAN（路径聚合网络）** 进行多尺度特征提取，而许多 YOLO 模型的**下采样因子是 32**。这样可以确保输入图像在网络中经过 **5 次下采样（每次 stride=2）** 时，不会产生小数或对齐问题。  

### **为什么输入尺寸要是 32 的倍数？**  
YOLO 模型的主干网络（如 CSPDarknet、EfficientNet 等）通常会经历 **5 层 stride=2 的下采样**，最终得到 **1/32 尺寸的特征图**。如果输入尺寸不是 32 的倍数，就可能会出现以下问题：  

1. **特征图尺寸不对齐**  
   - 例如输入 **613 × 613**，计算得到：  
     \[
     613 ÷ 32 = 19.15
     \]  
     由于结果不是整数，网络需要额外填充（padding），可能影响检测精度。  

2. **额外的填充与计算开销增加**  
   - 若输入尺寸不是 32 的倍数，模型可能会**自动填充**，增加计算量并影响推理效率。  

### **最佳实践：推荐使用 32 的倍数作为输入尺寸**  
为了保证最佳的推理效果，建议使用 **32 的倍数尺寸**，例如：  
✅ `640 × 640`  
✅ `512 × 512`  
✅ `800 × 800`  

这样可以确保网络的下采样过程对齐，避免不必要的填充，提高推理速度和精度。
